app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,598 @@
+  public void performHandshakeWithSelectedPeerDevice(@NonNull WifiP2pInfo groupInfo) {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Starting handshake");
+    }
+    peerGroupHandshakeAsyncTask = new PeerGroupHandshakeAsyncTask(this, groupInfo);
+    peerGroupHandshakeAsyncTask.execute();
+  }
+
+  /* Helper methods used for checking permissions and states of services */
+  private boolean checkCoarseLocationAccessPermission() { // Required by Android to detect wifi-p2p peers
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+
+      if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
+          != PackageManager.PERMISSION_GRANTED) {
+
+        if (shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_COARSE_LOCATION)) {
+          alertDialogShower.show(KiwixDialog.LocationPermissionRationale.INSTANCE,
+              new Function0<Unit>() {
+                @Override public Unit invoke() {
+                  requestPermissions(new String[] { Manifest.permission.ACCESS_COARSE_LOCATION },
+                      PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+                  return Unit.INSTANCE;
+                }
+              });
+        } else {
+          requestPermissions(new String[] { Manifest.permission.ACCESS_COARSE_LOCATION },
+              PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+        }
+
+        return false;
+      }
+    }
+
+    return true; // Control reaches here: Either permission granted at install time, or at the time of request
+  }
+
+  private boolean checkExternalStorageWritePermission() { // To access and store the zims
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+
+      if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+          != PackageManager.PERMISSION_GRANTED) {
+
+        if (shouldShowRequestPermissionRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+          alertDialogShower.show(KiwixDialog.StoragePermissionRationale.INSTANCE,
+              new Function0<Unit>() {
+                @Override public Unit invoke() {
+                  requestPermissions(new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+                      PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+                  return Unit.INSTANCE;
+                }
+              });
+        } else {
+          requestPermissions(new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+              PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+        }
+
+        return false;
+      }
+    }
+
+    return true; // Control reaches here: Either permission granted at install time, or at the time of request
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+      @NonNull int[] grantResults) {
+    switch (requestCode) {
+      case PERMISSION_REQUEST_CODE_COARSE_LOCATION: {
+        if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
```

Comment:
mildly confusing boolean statement, as a a rule of thumb avoid using `!=`. It is much easier to read as "if result is denied do this" than "if the result is not granted do this".
You can also move the switch statement into the if statement and stop yourself from doing the same if condition twice

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/30b18553cb8437886cd723c5679191c297136857)